clippy_lints/src/needless_borrowed_ref.rs

```diff
@@ -45,15 +46,20 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for NeedlessBorrowedRef {
         }
 
         if_let_chain! {[
-            // Pat is a pattern whose node
-            // is a binding which "involves" a immutable reference...
-            let PatKind::Binding(BindingMode::BindByRef(MutImmutable), ..) = pat.node,
-            // Pattern's type is a reference. Get the type and mutability of referenced value (tam: TypeAndMut).
-            let ty::TyRef(_, ref tam) = cx.tables.pat_ty(pat).sty,
-            // This is an immutable reference.
-            tam.mutbl == MutImmutable,
+            // Check immutable references.
+            // Only lint immutable refs, because `&mut ref T` may be useful.
+            let PatKind::Ref(ref sub_pat, MutImmutable) = pat.node,
+
+            // Check sub_pat got a 'ref' keyword.
+            let ty::TyRef(_, _) = cx.tables.pat_ty(sub_pat).sty,
         ], {
-            span_lint(cx, NEEDLESS_BORROWED_REFERENCE, pat.span, "this pattern takes a reference on something that is being de-referenced")
+            let part_to_keep = Span{ lo: pat.span.lo + BytePos(5), hi: pat.span.hi, ctxt: pat.span.ctxt };
```

Comment:
Please don't hand roll spans. It will bite us badly with macros and such. Can't you extract the span from the sub_Pat or a field of it?